{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d12b0c19",
   "metadata": {},
   "source": [
    "**Finding motifs with the highest frequency in the string?\"**\n",
    "\n",
    "Our aim is to turn the biological challenge of finding regulatory motifs into a computational problem. Below, we have implanted a 15-mer hidden message at a randomly selected position in each of ten randomly generated DNA strings. This example mimics a transcription factor binding site hiding in the upstream regions of ten genes.**\n",
    "\n",
    "1 \"atgaccgggatactgataaaaaaaagggggggggcgtacacattagataaacgtatgaagtacgttagactcggcgccgccg\"\n",
    "2 \"acccctattttttgagcagatttagtgacctggaaaaaaaatttgagtacaaaacttttccgaataaaaaaaaaggggggga\"\n",
    "3 \"tgagtatccctgggatgacttaaaaaaaagggggggtgctctcccgatttttgaatatgtaggatcattcgccagggtccga\"\n",
    "4 \"gctgagaattggatgaaaaaaaagggggggtccacgcaatcgcgaaccaacgcggacccaaaggcaagaccgataaaggaga\"\n",
    "5 \"tcccttttgcggtaatgtgccgggaggctggttacgtagggaagccctaacggacttaataaaaaaaagggggggcttatag\"\n",
    "6 \"gtcaatcatgttcttgtgaatggatttaaaaaaaaggggggggaccgcttggcgcacccaaattcagtgtgggcgagcgcaa\"\n",
    "7 \"cggttttggcccttgttagaggcccccgtaaaaaaaagggggggcaattatgagagagctaatctatcgcgtgcgtgttcat\"\n",
    "8 \"aacttgagttaaaaaaaagggggggctggggcacatacaagaggagtcttccttatcagttaatgctgtatgacactatgta\"\n",
    "9 \"ttggcccattggctaaaagcccaacttgacaaatggaagatagaatccttgcataaaaaaaagggggggaccgaaagggaag\"\n",
    "10 \"ctggtgagcaacgacagattcttacgtgcattagctcgcttccggggatctaatagcacgaagcttaaaaaaaaggggggga\"\n",
    "\n",
    "Can you find the implanted hidden message?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3cd73fdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The k-mer of length 15 with maximum frequency is:\n",
      "aaaaaaaaggggggg\n"
     ]
    }
   ],
   "source": [
    "def FrequentWords(Text, k):\n",
    "    Count = {}\n",
    "    for i in range(0, len(Text) - k):\n",
    "        Pattern = Text[i : k + i]\n",
    "        Count[Pattern] = Text.count(Pattern)\n",
    "\n",
    "    max_freq = sorted(set(Count.values()))[-1] # sort the dictionary values and -1 picks the largest value from among the list\n",
    "    \n",
    "    print(f\"The k-mer of length {k} with maximum frequency is:\")\n",
    "    for j,k in Count.items():\n",
    "        if k == max_freq:\n",
    "            print(j)\n",
    "\n",
    "dna = open('data.txt','r').read()\n",
    "FrequentWords(dna,15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc799d1c",
   "metadata": {},
   "source": [
    "**A brute force algorithm for motif finding:**\n",
    "\n",
    "Given a collection of strings Dna and an integer d, a k-mer is a (k,d)-motif if it appears in every string from Dna with at most d mismatches. For example, the implanted 15-mer in the strings above represents a (15,4)-motif.\n",
    "\n",
    "Implanted Motif Problem: Find all (k, d)-motifs in a collection of strings.\n",
    "\n",
    "Input: A collection of strings Dna, and integers k and d.\n",
    "Output: All (k, d)-motifs in Dna.\n",
    "\n",
    "**Note:**\n",
    "If a kmer of length k generated from one string has its neighbour matching with kmer generated from all other strings with a mismatch d, then the neighbour is considered a motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a79383fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def HammingDistance(p, q):\n",
    "    mismatch = 0\n",
    "    for i in range(len(p)):\n",
    "        if p[i] != q[i]:\n",
    "            mismatch += 1\n",
    "    return mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d0ec5f88",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Neighbors(Pattern, d):\n",
    "    if d == 0:\n",
    "        return {Pattern}\n",
    "    if len(Pattern) == 1:\n",
    "        return {'A', 'C', 'G', 'T'}\n",
    "\n",
    "    neighborhood = set()\n",
    "    suffix_neighbors = Neighbors(Pattern[1:], d)\n",
    "    for neighbor in suffix_neighbors:\n",
    "        if HammingDistance(Pattern[1:], neighbor) < d:\n",
    "            for nucleotide in {'A', 'C', 'G', 'T'}:\n",
    "                neighborhood.add(nucleotide + neighbor)\n",
    "        else:\n",
    "            neighborhood.add(Pattern[0] + neighbor)\n",
    "    return neighborhood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f96d3a2f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AAAAG',\n",
       " 'AAACG',\n",
       " 'AAACT',\n",
       " 'AAAGA',\n",
       " 'AAAGC',\n",
       " 'AAAGG',\n",
       " 'AAAGT',\n",
       " 'AAATC',\n",
       " 'AAATG',\n",
       " 'AACAT',\n",
       " 'AACGG',\n",
       " 'AACGT',\n",
       " 'AACTC',\n",
       " 'AACTG',\n",
       " 'AACTT',\n",
       " 'AAGAA',\n",
       " 'AAGAC',\n",
       " 'AAGAG',\n",
       " 'AAGAT',\n",
       " 'AAGCA',\n",
       " 'AAGCC',\n",
       " 'AAGCG',\n",
       " 'AAGCT',\n",
       " 'AAGGA',\n",
       " 'AAGGC',\n",
       " 'AAGGG',\n",
       " 'AAGGT',\n",
       " 'AAGTA',\n",
       " 'AAGTC',\n",
       " 'AAGTG',\n",
       " 'AAGTT',\n",
       " 'AATCG',\n",
       " 'AATGA',\n",
       " 'AATGC',\n",
       " 'AATGG',\n",
       " 'AATGT',\n",
       " 'AATTG',\n",
       " 'AATTT',\n",
       " 'ACAAG',\n",
       " 'ACAGG',\n",
       " 'ACAGT',\n",
       " 'ACATA',\n",
       " 'ACATC',\n",
       " 'ACATG',\n",
       " 'ACCGT',\n",
       " 'ACCTA',\n",
       " 'ACCTG',\n",
       " 'ACGAA',\n",
       " 'ACGAG',\n",
       " 'ACGCG',\n",
       " 'ACGCT',\n",
       " 'ACGGA',\n",
       " 'ACGGC',\n",
       " 'ACGGG',\n",
       " 'ACGGT',\n",
       " 'ACGTA',\n",
       " 'ACGTC',\n",
       " 'ACGTG',\n",
       " 'ACTAC',\n",
       " 'ACTAG',\n",
       " 'ACTGC',\n",
       " 'ACTGT',\n",
       " 'ACTTA',\n",
       " 'AGAAC',\n",
       " 'AGACG',\n",
       " 'AGAGC',\n",
       " 'AGAGG',\n",
       " 'AGAGT',\n",
       " 'AGATA',\n",
       " 'AGATC',\n",
       " 'AGATG',\n",
       " 'AGATT',\n",
       " 'AGCAA',\n",
       " 'AGCAC',\n",
       " 'AGCAG',\n",
       " 'AGCAT',\n",
       " 'AGCCG',\n",
       " 'AGCGC',\n",
       " 'AGCGG',\n",
       " 'AGCTA',\n",
       " 'AGCTG',\n",
       " 'AGCTT',\n",
       " 'AGGAA',\n",
       " 'AGGAC',\n",
       " 'AGGAG',\n",
       " 'AGGAT',\n",
       " 'AGGCA',\n",
       " 'AGGCG',\n",
       " 'AGGCT',\n",
       " 'AGGGA',\n",
       " 'AGGGC',\n",
       " 'AGGGG',\n",
       " 'AGGGT',\n",
       " 'AGGTA',\n",
       " 'AGGTC',\n",
       " 'AGGTG',\n",
       " 'AGGTT',\n",
       " 'AGTAA',\n",
       " 'AGTAC',\n",
       " 'AGTAG',\n",
       " 'AGTCA',\n",
       " 'AGTCC',\n",
       " 'AGTCG',\n",
       " 'AGTCT',\n",
       " 'AGTGA',\n",
       " 'AGTGC',\n",
       " 'AGTGG',\n",
       " 'AGTGT',\n",
       " 'AGTTA',\n",
       " 'AGTTC',\n",
       " 'AGTTG',\n",
       " 'AGTTT',\n",
       " 'ATAAG',\n",
       " 'ATACG',\n",
       " 'ATACT',\n",
       " 'ATAGG',\n",
       " 'ATAGT',\n",
       " 'ATATA',\n",
       " 'ATATG',\n",
       " 'ATCAG',\n",
       " 'ATCGA',\n",
       " 'ATCGT',\n",
       " 'ATCTA',\n",
       " 'ATCTG',\n",
       " 'ATCTT',\n",
       " 'ATGAG',\n",
       " 'ATGAT',\n",
       " 'ATGCC',\n",
       " 'ATGCG',\n",
       " 'ATGCT',\n",
       " 'ATGGA',\n",
       " 'ATGGC',\n",
       " 'ATGGT',\n",
       " 'ATGTA',\n",
       " 'ATTAG',\n",
       " 'ATTGC',\n",
       " 'ATTTA',\n",
       " 'CAAAG',\n",
       " 'CAAAT',\n",
       " 'CAACG',\n",
       " 'CAACT',\n",
       " 'CAAGA',\n",
       " 'CAAGC',\n",
       " 'CAAGG',\n",
       " 'CAAGT',\n",
       " 'CAATG',\n",
       " 'CAATT',\n",
       " 'CACCT',\n",
       " 'CAGAT',\n",
       " 'CAGCA',\n",
       " 'CAGCT',\n",
       " 'CAGGA',\n",
       " 'CAGGG',\n",
       " 'CAGGT',\n",
       " 'CAGTC',\n",
       " 'CAGTG',\n",
       " 'CAGTT',\n",
       " 'CATAC',\n",
       " 'CATAG',\n",
       " 'CATAT',\n",
       " 'CATCA',\n",
       " 'CATCC',\n",
       " 'CATCT',\n",
       " 'CATGA',\n",
       " 'CATGC',\n",
       " 'CATTC',\n",
       " 'CCAAG',\n",
       " 'CCACG',\n",
       " 'CCAGG',\n",
       " 'CCAGT',\n",
       " 'CCGAG',\n",
       " 'CCGAT',\n",
       " 'CCGCA',\n",
       " 'CCGCT',\n",
       " 'CCGGA',\n",
       " 'CCGGC',\n",
       " 'CCGGG',\n",
       " 'CCGGT',\n",
       " 'CCGTG',\n",
       " 'CCTAT',\n",
       " 'CCTGA',\n",
       " 'CCTGC',\n",
       " 'CGAAC',\n",
       " 'CGACG',\n",
       " 'CGAGA',\n",
       " 'CGAGG',\n",
       " 'CGAGT',\n",
       " 'CGATA',\n",
       " 'CGATC',\n",
       " 'CGATG',\n",
       " 'CGATT',\n",
       " 'CGCAA',\n",
       " 'CGCAG',\n",
       " 'CGCGA',\n",
       " 'CGCTA',\n",
       " 'CGGAA',\n",
       " 'CGGAC',\n",
       " 'CGGAG',\n",
       " 'CGGAT',\n",
       " 'CGGCA',\n",
       " 'CGGCC',\n",
       " 'CGGCT',\n",
       " 'CGGGA',\n",
       " 'CGGGC',\n",
       " 'CGGGG',\n",
       " 'CGGGT',\n",
       " 'CGGTA',\n",
       " 'CGGTC',\n",
       " 'CGGTG',\n",
       " 'CGGTT',\n",
       " 'CGTAA',\n",
       " 'CGTAC',\n",
       " 'CGTAT',\n",
       " 'CGTCA',\n",
       " 'CGTCC',\n",
       " 'CGTCT',\n",
       " 'CGTGA',\n",
       " 'CGTGC',\n",
       " 'CGTGG',\n",
       " 'CGTGT',\n",
       " 'CGTTA',\n",
       " 'CGTTC',\n",
       " 'CTAAG',\n",
       " 'CTACA',\n",
       " 'CTACC',\n",
       " 'CTACT',\n",
       " 'CTAGG',\n",
       " 'CTAGT',\n",
       " 'CTCAG',\n",
       " 'CTCAT',\n",
       " 'CTCGA',\n",
       " 'CTCGC',\n",
       " 'CTGAA',\n",
       " 'CTGAG',\n",
       " 'CTGAT',\n",
       " 'CTGCG',\n",
       " 'CTGCT',\n",
       " 'CTGGA',\n",
       " 'CTGGC',\n",
       " 'CTGGG',\n",
       " 'CTGTC',\n",
       " 'CTGTG',\n",
       " 'CTTAC',\n",
       " 'CTTAT',\n",
       " 'CTTCG',\n",
       " 'CTTCT',\n",
       " 'CTTGA',\n",
       " 'CTTGC',\n",
       " 'CTTGG',\n",
       " 'GAAAA',\n",
       " 'GAAAC',\n",
       " 'GAAAG',\n",
       " 'GAAAT',\n",
       " 'GAACG',\n",
       " 'GAACT',\n",
       " 'GAAGA',\n",
       " 'GAAGC',\n",
       " 'GAAGG',\n",
       " 'GAAGT',\n",
       " 'GAATC',\n",
       " 'GAATG',\n",
       " 'GAATT',\n",
       " 'GACAC',\n",
       " 'GACAT',\n",
       " 'GACCG',\n",
       " 'GACGG',\n",
       " 'GACGT',\n",
       " 'GACTA',\n",
       " 'GACTG',\n",
       " 'GACTT',\n",
       " 'GAGAA',\n",
       " 'GAGAC',\n",
       " 'GAGAG',\n",
       " 'GAGAT',\n",
       " 'GAGCG',\n",
       " 'GAGGA',\n",
       " 'GAGGC',\n",
       " 'GAGGG',\n",
       " 'GAGGT',\n",
       " 'GAGTA',\n",
       " 'GAGTC',\n",
       " 'GAGTG',\n",
       " 'GAGTT',\n",
       " 'GATAA',\n",
       " 'GATAC',\n",
       " 'GATAG',\n",
       " 'GATAT',\n",
       " 'GATCA',\n",
       " 'GATCG',\n",
       " 'GATCT',\n",
       " 'GATGA',\n",
       " 'GATGC',\n",
       " 'GATGG',\n",
       " 'GATTA',\n",
       " 'GATTC',\n",
       " 'GATTG',\n",
       " 'GATTT',\n",
       " 'GCAAA',\n",
       " 'GCAAC',\n",
       " 'GCAAG',\n",
       " 'GCAAT',\n",
       " 'GCACA',\n",
       " 'GCACG',\n",
       " 'GCACT',\n",
       " 'GCAGA',\n",
       " 'GCATA',\n",
       " 'GCATC',\n",
       " 'GCATG',\n",
       " 'GCCAC',\n",
       " 'GCCAG',\n",
       " 'GCCAT',\n",
       " 'GCCCA',\n",
       " 'GCCCG',\n",
       " 'GCCGA',\n",
       " 'GCCGG',\n",
       " 'GCGAA',\n",
       " 'GCGAC',\n",
       " 'GCGAG',\n",
       " 'GCGCA',\n",
       " 'GCGGA',\n",
       " 'GCGGC',\n",
       " 'GCGGG',\n",
       " 'GCGGT',\n",
       " 'GCGTC',\n",
       " 'GCGTG',\n",
       " 'GCGTT',\n",
       " 'GCTAA',\n",
       " 'GCTAC',\n",
       " 'GCTAG',\n",
       " 'GCTAT',\n",
       " 'GCTCA',\n",
       " 'GCTCC',\n",
       " 'GCTCG',\n",
       " 'GCTGC',\n",
       " 'GCTGG',\n",
       " 'GCTGT',\n",
       " 'GCTTA',\n",
       " 'GCTTC',\n",
       " 'GCTTG',\n",
       " 'GCTTT',\n",
       " 'GGAAA',\n",
       " 'GGAAC',\n",
       " 'GGAAG',\n",
       " 'GGACG',\n",
       " 'GGAGC',\n",
       " 'GGAGG',\n",
       " 'GGAGT',\n",
       " 'GGATA',\n",
       " 'GGATT',\n",
       " 'GGCAA',\n",
       " 'GGCAC',\n",
       " 'GGCAG',\n",
       " 'GGCAT',\n",
       " 'GGCCA',\n",
       " 'GGCCG',\n",
       " 'GGCGA',\n",
       " 'GGCGG',\n",
       " 'GGCTA',\n",
       " 'GGCTC',\n",
       " 'GGCTG',\n",
       " 'GGCTT',\n",
       " 'GGGAA',\n",
       " 'GGGAC',\n",
       " 'GGGAG',\n",
       " 'GGGAT',\n",
       " 'GGGCA',\n",
       " 'GGGCT',\n",
       " 'GGGGA',\n",
       " 'GGGGT',\n",
       " 'GGGTA',\n",
       " 'GGGTC',\n",
       " 'GGGTG',\n",
       " 'GGGTT',\n",
       " 'GGTAA',\n",
       " 'GGTAC',\n",
       " 'GGTAG',\n",
       " 'GGTAT',\n",
       " 'GGTCA',\n",
       " 'GGTCG',\n",
       " 'GGTGA',\n",
       " 'GGTGC',\n",
       " 'GGTGG',\n",
       " 'GGTGT',\n",
       " 'GGTTA',\n",
       " 'GGTTC',\n",
       " 'GGTTG',\n",
       " 'GGTTT',\n",
       " 'GTAAA',\n",
       " 'GTAAC',\n",
       " 'GTAAG',\n",
       " 'GTAAT',\n",
       " 'GTACC',\n",
       " 'GTACG',\n",
       " 'GTACT',\n",
       " 'GTAGA',\n",
       " 'GTAGC',\n",
       " 'GTAGG',\n",
       " 'GTATG',\n",
       " 'GTCAA',\n",
       " 'GTCAC',\n",
       " 'GTCAG',\n",
       " 'GTCCA',\n",
       " 'GTCCT',\n",
       " 'GTCGA',\n",
       " 'GTCGG',\n",
       " 'GTCGT',\n",
       " 'GTCTA',\n",
       " 'GTGAA',\n",
       " 'GTGAC',\n",
       " 'GTGAG',\n",
       " 'GTGAT',\n",
       " 'GTGCA',\n",
       " 'GTGCC',\n",
       " 'GTGCG',\n",
       " 'GTGCT',\n",
       " 'GTGGA',\n",
       " 'GTGGC',\n",
       " 'GTGGG',\n",
       " 'GTGGT',\n",
       " 'GTGTC',\n",
       " 'GTGTG',\n",
       " 'GTGTT',\n",
       " 'GTTAA',\n",
       " 'GTTAC',\n",
       " 'GTTAG',\n",
       " 'GTTAT',\n",
       " 'GTTCA',\n",
       " 'GTTCC',\n",
       " 'GTTGA',\n",
       " 'GTTGT',\n",
       " 'GTTTA',\n",
       " 'TAAAG',\n",
       " 'TAAAT',\n",
       " 'TAACG',\n",
       " 'TAACT',\n",
       " 'TAAGA',\n",
       " 'TAAGC',\n",
       " 'TAAGG',\n",
       " 'TAAGT',\n",
       " 'TAATG',\n",
       " 'TAATT',\n",
       " 'TACAA',\n",
       " 'TACAC',\n",
       " 'TACAG',\n",
       " 'TACCA',\n",
       " 'TACCT',\n",
       " 'TACGA',\n",
       " 'TACGC',\n",
       " 'TACTA',\n",
       " 'TACTG',\n",
       " 'TAGAG',\n",
       " 'TAGAT',\n",
       " 'TAGCA',\n",
       " 'TAGCT',\n",
       " 'TAGGA',\n",
       " 'TAGGC',\n",
       " 'TAGGT',\n",
       " 'TAGTA',\n",
       " 'TAGTC',\n",
       " 'TAGTG',\n",
       " 'TAGTT',\n",
       " 'TATAC',\n",
       " 'TATGA',\n",
       " 'TATGC',\n",
       " 'TATTT',\n",
       " 'TCAAG',\n",
       " 'TCAAT',\n",
       " 'TCAGA',\n",
       " 'TCAGC',\n",
       " 'TCAGT',\n",
       " 'TCATA',\n",
       " 'TCATC',\n",
       " 'TCCAA',\n",
       " 'TCCAG',\n",
       " 'TCCGA',\n",
       " 'TCGAA',\n",
       " 'TCGAC',\n",
       " 'TCGAG',\n",
       " 'TCGAT',\n",
       " 'TCGGG',\n",
       " 'TCGGT',\n",
       " 'TCGTG',\n",
       " 'TCTAC',\n",
       " 'TCTAG',\n",
       " 'TCTGA',\n",
       " 'TGAAA',\n",
       " 'TGAAC',\n",
       " 'TGAAG',\n",
       " 'TGACT',\n",
       " 'TGAGA',\n",
       " 'TGAGG',\n",
       " 'TGAGT',\n",
       " 'TGATA',\n",
       " 'TGATC',\n",
       " 'TGATG',\n",
       " 'TGATT',\n",
       " 'TGCAA',\n",
       " 'TGCCA',\n",
       " 'TGCGA',\n",
       " 'TGCTA',\n",
       " 'TGCTC',\n",
       " 'TGCTG',\n",
       " 'TGCTT',\n",
       " 'TGGAG',\n",
       " 'TGGCA',\n",
       " 'TGGCC',\n",
       " 'TGGCG',\n",
       " 'TGGCT',\n",
       " 'TGGGA',\n",
       " 'TGGGC',\n",
       " 'TGGGG',\n",
       " 'TGGGT',\n",
       " 'TGGTA',\n",
       " 'TGGTG',\n",
       " 'TGTAA',\n",
       " 'TGTGA',\n",
       " 'TGTTA',\n",
       " 'TTAAG',\n",
       " 'TTAAT',\n",
       " 'TTACA',\n",
       " 'TTAGT',\n",
       " 'TTATG',\n",
       " 'TTCCG',\n",
       " 'TTCGA',\n",
       " 'TTGAT',\n",
       " 'TTGCG',\n",
       " 'TTGCT',\n",
       " 'TTGGA',\n",
       " 'TTGGC'}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def MotifEnumeration(Dna, k, d):\n",
    "    Patterns = []\n",
    "    potential_motif = []\n",
    "    strings = Dna.split()\n",
    "    for i in strings:\n",
    "        for j in range(len(i) - k + 1):\n",
    "            k_mer = i[j:j+k]\n",
    "            neighbourhood = Neighbors(k_mer, d)\n",
    "            for neighbour in neighbourhood:\n",
    "                for m in strings:\n",
    "                    if m != i:\n",
    "                        for l in range(len(m) - k + 1):\n",
    "                            mismatch = HammingDistance(neighbour, m[l:l+k])\n",
    "                            if mismatch <= d:\n",
    "                                potential_motif.append(m)\n",
    "                if len(set(potential_motif)) == len(strings) - 1:\n",
    "                    Patterns.append(neighbour)\n",
    "                potential_motif = []\n",
    "    return set(Patterns)\n",
    "\n",
    "MotifEnumeration('GCATCCAAGAAGACTGGTAAATGCA GAAGGTCGGGTACGTGTGCATGTGC GCGGCGAGAGTAATAAAAGCTTACA AGACAGGTGAGGCACTTGTTCAAGC CGAGCTAAGAGGGCGAGACCGATCT TTCATTTTGCTAAGTGGTTCGGTAC', 5, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d764589",
   "metadata": {},
   "source": [
    "**Compute the entropy of the NF-kB matrix**\n",
    "\n",
    "Entropy = -p . log 2 (p)\n",
    "\n",
    "Entropy is an indication of conservation calculated from the entropy of each column of the matrix. The more conserved the column, the smaller its entropy.\n",
    "\n",
    "Entropy is a better metric to score motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1f6b05fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9.916290005356972"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "def entropy(strings):\n",
    "    dna = strings.split()\n",
    "    num_dna = len(dna)\n",
    "    entropy_cal = 0\n",
    "    a, t, g, c = 0, 0, 0, 0\n",
    "    for i in range(0,len(dna[0])):\n",
    "        for j in dna:\n",
    "            if j[i] == 'A':\n",
    "                a += 1\n",
    "            elif j[i] == 'T':\n",
    "                t += 1\n",
    "            elif j[i] == 'G':\n",
    "                g += 1\n",
    "            else:\n",
    "                c += 1\n",
    "                \n",
    "        a /= num_dna\n",
    "        t /= num_dna\n",
    "        g /= num_dna\n",
    "        c /= num_dna\n",
    "        if a > 0:\n",
    "            entropy_cal += -(a * math.log2(a))\n",
    "        if t > 0:\n",
    "            entropy_cal += -(t * math.log2(t))\n",
    "        if g > 0:\n",
    "            entropy_cal += -(g * math.log2(g))\n",
    "        if c > 0:\n",
    "            entropy_cal += -(c * math.log2(c))\n",
    "        a, t, g, c = 0, 0, 0, 0\n",
    "    \n",
    "    return entropy_cal\n",
    "    \n",
    "entropy(\"TCGGGGGTTTTT CCGGTGACTTAC ACGGGGATTTTC TTGGGGACTTTT AAGGGGACTTCC TTGGGGACTTCC TCGGGGATTCAT TCGGGGATTCCT TAGGGGAACTAC TCGGGTATAACC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d0db85d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "67"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def DistanceBetweenPatternAndStrings(Pattern, Dna):\n",
    "    k = len(Pattern)\n",
    "    distance = sum(min(HammingDistance(Pattern, text[i:i+k]) for i in range(len(text) - k + 1)) for text in Dna.split())\n",
    "    return distance\n",
    "\n",
    "DistanceBetweenPatternAndStrings(\"ATGATTA\", \"CCACGTCGCGAACCATACCCGCCATCGTTCGTGCGGTTCCTAGTACAGGGACACTGACGAAACTAATATAGGAGCGAGTAGTCTTGATATTCCATTGACCAAT AGATTTGGTTGTTTCGTGGTGTCGAATTGGTCGCAGATGATTCGCCTGGTAGGCGATTTCACCCTTGCCCCGTTTAGGGAAATACTAGCCACTACACTGATCT TAGACCAATATTGTAAATTGCGCTGCGGCAGTGCGATTATCGGCGATGCAAGCAAGGGGGCAACTCTCATGAGCATATTTGCTTCGCAAGCCTCCAGCAGTAT GGGGCCCAGCTGATGGCTGTGGGTGTGTGGCATTAAACAGTATGTACTTAATGACAGGTGCGTAAACGCATCATAGGTTTACGTTACGACGTTCTCTGCGAGA GATGTTAGCATATTCCCACCAGCCAGACCACAGAATACGGAGACCACGGTGCTCCCTCCATGAACCCGAGTGCTAGGTGTACGACAACAAAACACTTAAGGCA CCTAGAGCGTTAGAAGAGCAAGCGAAATTTTAAAAAATACGCTTTGTCAACAAACTGCGCAACAGCATATGGCTGTGCGGGAGTAGGAAGTGATATGCGTTTC CAATCATATCCAGATAGTGGTATCAGCAGTGGTGCAATGGGCAGAGTCAGGCATAGATCGGCGCGTAAATACTTATCGTGCCTAGGTTGACGATCCATAGCTA GACTCCACAATGAACAGCGAGCTTGGGTCTATTTGGCGGTCTAACCTTAATGTTTTCTGAACGCTTTTTTTACTCTTCACAGAGCATACAAACCACCGATAGA TACCATTCGAGATGAATGTCAGCCAAAATCCCGTCTGTCCAAAGTCCCGTGACTTATCGCGCAGAATTCAAGCTGTTTATGACCGACGATACAGATCATCTGA GATGATTTATTGGCGATCCTAGTGAATGTGTAGTAGGGGCAGCTTTAGGGAAGTCAAATTCCTAGAGTTCGAAGTTCTGTCTACAAAAATCAGATGTCGCTCG TCAGTTACGTACCGCAGGTCAGCTGAGATGTAATAGGCTCCACCTGTTACGGGCACTCTGGATCGGTGTAAGAATTGGGTACACAGGTAACGGCGAGTGTAGA TCGAGAACCCAGACTGACTTGAGCTGACGGCTTCGGTACAGATGAATTACGGTATCACGTCTATGCTGTCGCCATTAGGCTCAAACCGTAAAGAAGACATGTG TCGTGGTATCATCAAAGACTAATTCTCGACAGGACAAACGCGACTGGCGGATAATAGGATACGATTGGCGATGTGCCTACGTGGATTCGGCGGGGTCAGGCAT GTCGGGATCCTTATCTTGGCATTTCATGCCCCAGAATGCATAGCCATTCACCTTAGTAAAAATGCCCTTACGTTGCACTCTTTCCAGGCGCACCGCTGCGCTC GAGCAGTTGTAAGCCAATTACTTGGGAGAGTAGGCTAGTGATAGCCGCACATGCAGAACCGATTTGGCCGGTAACCTATTACCAGACATCTGACTCGTATGAA GACCGCCCCAAGTCCGTGGACTTTAGCTCGGATCCGTACACCTTAGCGTCACTTCACGCCCTTCCGACCGCGAGGGTAGATTACCTGGCGCCTAGATAATCAC TAACATCCCCGCGTGTTGGGACGATGTTCCATTTATGGGAAAGGCACAAGAAATCTGAAAGTAACTAGGGTAATAGATCGTCGTGTCATACAATTTTTAACGT TGCCAGGGTAACTAGAGCACCGTCTAGGCAACGTCAAACGTGACAGGTGATGCGCCGCTTCTACGAAGTGTCTAGCCGACACTTCGGAATAACGACTGCACTG GGATTTAACGTACAGGAAACCGTCTCTGCTGCCAAGGCCCCACAAACTACTCAGAGAGAAAGTGTCGTGGCCAGCAGAGAGGCGTCGGTTGCATGCGAATCGG GGAATGTAACATAATGCCGTACCCCCAGGGATTATTTGTGAAGCATTTCTACTCACGAATCGGCCTTTAGCCCGAAGCAACGTATGGTCATACTTTGCTAAAA TTTAAGTCACTCGTTACCTATGACTTCTGGTCGTATATCTATATTACCTGTCGATGAAGATTGCAAGAGTGGGCAATTCTTACGTTCTCAATGGAGATATATT GTACTAGTCACTGCTGGCATATTCCAGAGGTTTGGGTTTAAACAGTCCAGCGGGGGTCTCCACTTGCCCGTAAAACCGCCCAATCGATGTCATTTTGGGATTG TGGTTCTTTCGGCTCGGAATCCTATCTTTATAAAAATAATGGCAGACCGTTTACGTTGCTAGCTGTTTTCTCCACTCAGAAATGTGGACAGCAGACGGGAGTT TTGGGTTGTGGAGAGCAATGGTTCAATAGTTTCTGGACGTTGCGAATGGTAATATGTATCACGACCGTCACGAATCTTTCACGGCATGATGCGATCGTGTTAA TATTCAAGCCCTATCCCCGGAACCATTTTCCACGTGGTATCAAAGGTCTTGTGCCTCCGGACACCATATGCGGCACATGGTTTTCCAGCTCAGATTGTTCAAG CCACCTGTACCCAGAGTATGCGAAACCTGACGGTCGCCCAGGTTCGGTTTCAAGGAGACCTGGCTGACGGATTTACAAAGGCTGTTTTTCTACTATGGTACGG GGACACATTTGACGGTCTAGTCGCGCCATGCCTCACTAACGCTGTTCGAATGTTCTCAATTGATTCGATTTCACGCGAGCAGATGTCCCTGCGTAACAGTCAG CAAAGTGGCCATAAACCTGATTGAGGCACTGTCCTCCGTAGGGGGGTCTATTGATCTGTCCGACCGTAAAGACAAGAAAGTCCTGTGTGAGACAACCGACAAA TGTTAATACCGAATGCCTACGGCGGCTCGGAGGACGAAATTAAGACCTCTGGTTGCCTCGAGATTGTGTTGTCTACCGCGGCTCAATACAGGCAAATCCCGGT GACTACTCTCATGTGGTAACCTTCATTATAGGTGCTACTGGTTGACGGGTGAGATGACGAATTGATAACGGCAGTGTCTGACACAGTAGTATTGAACAACGTC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "09c4fb19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGGGCT'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def MedianString(k, Dna):\n",
    "    alphabet = ['A', 'C', 'G', 'T']\n",
    "    distance = float('inf')\n",
    "    median = ''\n",
    "\n",
    "    for i in range(4 ** k):\n",
    "        pattern = ''\n",
    "        index = i\n",
    "        for _ in range(k):\n",
    "            pattern = alphabet[index % 4] + pattern\n",
    "            '''\n",
    "            Since there are four possible nucleotides (A, C, G, T), using remainder division by 4 ensures that the \n",
    "            remainder will be in the range [0, 1, 2, 3]\n",
    "            '''\n",
    "            index //= 4\n",
    "            '''\n",
    "            After extracting the nucleotide, index needs to be updated to consider the next bits for the next iteration\n",
    "            '''\n",
    "            \n",
    "        if distance > DistanceBetweenPatternAndStrings(pattern, Dna):\n",
    "            distance = DistanceBetweenPatternAndStrings(pattern, Dna)\n",
    "            median = pattern\n",
    "\n",
    "    return median\n",
    "\n",
    "MedianString(6, \"TGCCGGAATAGTCCTCGTTCCTGAAGTGCTAGCTAGAAGCTA CACGTAGCTATGCTGTAGCAGGACCGCCAAAGCGCTTCGTGC TTGGAGGTCTGTTAATTCAGGGCTGACCACGTCTACCGTAGG CAGATTACGTAAAGGGCTATCCTGTAGGACACCGGTCCATAA ACGTCTATCTGCAGGGAGTTTTCCAGCGCTGTCAGGGGGGCG CGGCTGTGTGTAGGTAGCAACGGTTGGATTACATAGAGCGCT ATGCTGAGGGCTTGCAATACAGGCACATCAACGATTGGTGTT AGCGTAGACGCGAGAGCTCACCAAAGCACGATCCGCCTAACC GTCCCGGTAATTTACACCCCCGTAAGTGCTCCGTATGAGAAC GTTGCACTGCACGCAACACCTCCTGAAATCTATCGGAGGGCT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0a51a709",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGTTTT'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def median_string(k, dna):\n",
    "    distance = float('inf')\n",
    "    pattern = \"\"\n",
    "\n",
    "    for i in range(len(dna[0]) - k + 1):\n",
    "        current_pattern = dna[0][i:i+k]\n",
    "        current_distance = sum(min(HammingDistance(current_pattern, seq[i:i+k]) for seq in dna) for i in range(len(dna[0]) - k + 1))\n",
    "\n",
    "        if current_distance < distance:\n",
    "            distance = current_distance\n",
    "            pattern = current_pattern\n",
    "\n",
    "    return pattern\n",
    "\n",
    "median_string(6, [\"AAGCTTTCCTTTACAGTTTTCCAGCACAGCCGGAGGTGAATG\", \"GAAAGGACAGTTAACGTGACTGACGCGCTCCGGAATATTAGA\", \"ATCGGGTAGGTTGAGGTGAAAGTTCTAAACAAAACGTAGACG\", \"GGACCATCACGGCCTGGATGGCGAACTGACAGAGTTGCTTTT\", \"TGACCGAGTTCTTAGAGGTTCCGGACAGTGAACCCCAGAGTT\", \"GAAGTGAGAAAGGACGTTATAGTTCCTAAATTCAGAACCATG\", \"ACAGTTAAGTGTCCTCGGGGTTTTGTACACTGAGTCGTTGGA\", \"ACATAAGTAGCGGAGTGAATGACGTAGCAGATAGTTAGGGTT\", \"TAAATCAATTTAGTTCCGATAGTTGGCTGTTTCGCAAAATCA\", \"CGCCCGCCATCAAACTCTTCGGGCTTACCGTTTGGGAAAGTT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "31c79233",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Most probable kmer is TCCGGAGATTTT with score 3.2036092294833274e-07\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TCCGGAGATTTT'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def Profile_MostProbable_kmer(string, k, profile):\n",
    "    kmer_score = {}\n",
    "    for i in range(len(string) - k + 1):\n",
    "        kmer = string[i:i+k]\n",
    "        score = 1\n",
    "        for j in range(len(kmer)):\n",
    "            score *= profile.loc[kmer[j], j]\n",
    "        kmer_score[kmer] = score\n",
    "    \n",
    "    max_score = max(kmer_score.values())\n",
    "    print(f\"Most probable kmer is {list(kmer_score.keys())[list(kmer_score.values()).index(max_score)]} with score {max_score}\")\n",
    "    return list(kmer_score.keys())[list(kmer_score.values()).index(max_score)]\n",
    "            \n",
    "data = [[0.277, 0.265, 0.265, 0.229, 0.217, 0.313, 0.229, 0.313, 0.277, 0.181, 0.193, 0.229], [0.265, 0.301, 0.265, 0.253, 0.265, 0.193, 0.157, 0.253, 0.277, 0.229, 0.337, 0.217], [0.217, 0.241, 0.277, 0.325, 0.289, 0.229, 0.313, 0.181, 0.205, 0.289, 0.205, 0.253], [0.241, 0.193, 0.193, 0.193, 0.229, 0.265, 0.301, 0.253, 0.241, 0.301, 0.265, 0.301]]\n",
    "df = pd.DataFrame(data, index = ['A', 'C', 'G', 'T'])\n",
    "Profile_MostProbable_kmer('TTGGACATAACGGTAGTGCATAACCGTAATACTTTATAACCTTCAGATCCACCACAACAGCAACCGCACTTATGCTCCCAGTCACCAATGATAGCAAGACGGCGCTCGCCTCTATGAAATGTCTCGCGAGTGCCTATTAGAGGCTCTTACAGTACACAGGTCAGACGAAAGGTCCAAGTATTCCGTGTCCCTACCAAATGACCCTGGTTTAGGCTGAAACTCAGGCCTAATAGCTCACGCTGTAACGTTAGCCTTGCCAAGGATCATAAGGAGTGCCCGCATGCCGGGTCAGCCTATACAATTGAATCTAACGGCATACCTCTAGATTTCAAAGAAGGGGACGGGTCCCGTCAAAGTAAAAGAATACGAGGGATCGTCACTATATTGCTTTAAGAGGTTCTTGGTAGGTACGAATTCTTCGACTATAAAGAGAGCCTCACCGCTAACAACCATCGAAAGAGAAACATAAAATTGATCGGCCGACAGAATGACATCCAAACAGTGCACGGATGCCAAGGGTGCTTGGGACCTCGAGCTGAAAGAGCCTGAACGTATTATCTGACAGCTGAGTCGATTCATATGAGAGACCTCAACTAAGTGTCGCTTAGATAGCAACCGAATAGGCCCGTGCAAGTGGGCTTAACTATCTGAGGAGATCTGTAGGACTTAGGAAAGCGGGCGTGGGTGCTGCTGTCTAGCCCGATGAATCGTAACGAGATCTGAGTTATGCAAGAGAATGTTGTTCAGCTCCTTTGGGGTCTCACTTCTGAATCCGGTAATTGTTTCTGACCCCCTGATATAACAAGATCTTATACATTTTTAACAGCTACCTAATCGCGAATCAGACCCCTGCTCCGGTTCGGTTTGCAACTGTGGCTTCCCAAGGGAGACGTCCGGAGATTTTGATCTGCGGAATAAGCCCAGATCCGAATAACATAGCGCACGCCTCCCTCTTACTAACCGAGTTCCATATGCACAGACGTTTTACGTGCGTTCTCAC', 12, df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4683eaf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    nucleotides = {'A', 'C', 'G', 'T'}\n",
    "    max_probability = -1.0\n",
    "    most_probable_kmer = \"\"\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i+k]\n",
    "        probability = 1.0\n",
    "        for j, nucleotide in enumerate(kmer):\n",
    "            probability *= profile[nucleotide][j]\n",
    "        if probability > max_probability:\n",
    "            max_probability = probability\n",
    "            most_probable_kmer = kmer\n",
    "    return most_probable_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3bcc3e75",
   "metadata": {},
   "outputs": [],
   "source": [
    "def form_profile(motifs):\n",
    "    k = len(motifs[0])\n",
    "    profile = {nucleotide: [0] * k for nucleotide in 'ACGT'}\n",
    "    t = len(motifs)\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        for nucleotide in 'ACGT':\n",
    "            count = column.count(nucleotide)\n",
    "            profile[nucleotide][i] = count / t\n",
    "    return profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "c93bda3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def score_motifs(motifs):\n",
    "    k = len(motifs[0])\n",
    "    t = len(motifs)\n",
    "    score = 0\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        max_count = max(column.count(nucleotide) for nucleotide in 'ACGT')\n",
    "        score += t - max_count\n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "cb14b0db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTCAGTCTACGG\n",
      "GTTTACAGACGA\n",
      "CATCTAACGTCC\n",
      "TGGCCCCAACCT\n",
      "TTCATTCTACGC\n",
      "TTCACTCAACCA\n",
      "TACCCACAGCGC\n",
      "TTTATTCAATCC\n",
      "TTCAATCAACGA\n",
      "TTCAGTCTACCG\n",
      "GTGCTTAGGCCC\n",
      "TTCAATCCACGG\n",
      "TTCAGTCGACCC\n",
      "TGCTATCCACCT\n",
      "TTCAGTCTACCC\n",
      "TGCCTTAAGTCT\n",
      "TTCATTCAACGT\n",
      "TTCCATCAGCCG\n",
      "CGGGAGGATCTT\n",
      "CTGACACAACTT\n",
      "TTCAGTCTACTC\n",
      "TTCACTCTACTT\n",
      "TTCACTCAACGT\n",
      "TTCAGTCAACCT\n",
      "TTCAGTCTACTA\n"
     ]
    }
   ],
   "source": [
    "def greedy_motif_search(dna, k, t):\n",
    "    best_motifs = [seq[:k] for seq in dna]\n",
    "    for i in range(len(dna[0]) - k + 1):\n",
    "        motif1 = dna[0][i:i+k]\n",
    "        motifs = [motif1]\n",
    "        for j in range(1, t):\n",
    "            profile = form_profile(motifs)\n",
    "            most_probable_kmer = profile_most_probable_kmer(dna[j], k, profile)\n",
    "            motifs.append(most_probable_kmer)\n",
    "        if score_motifs(motifs) < score_motifs(best_motifs):\n",
    "            best_motifs = motifs\n",
    "    return best_motifs\n",
    "\n",
    "Dna = \"GATTCAGTCTACGGATAGCAATTGCTCGATGGATTTTACGCCGCACAATTGACTATGGATGGCTCTGTTGCATTTTGATATTCTTGAGCCATCACGAACTCTTTGCGACTGGCAATTCAATCAGTCTGCCCGTGTACAAATTAGTCAGCGCGGTGA GTTTACAGACGATGAATTAATCCCCAGCTCTTACTCCCGAGTAGAGAATAAGTGCCTCCACTATGGTATACAAGTTGTTAGGTGCCTTCAATCTACTAGACCTATGATGGCTCTTCGTCAAATAAACTAACAAGGAAGGTCCTTGCAGTTACTAGA CATCTAACGTCCCATTCACTCTACTGATCGTTCAAAACTGACTGTCGAGCCACCTACTGGTGCTCTCGGCTGTTGCTAACATTACCGTGCACATGCACCTTAGCGGTCGTCGATCAGAATGATTTGTCCTCCCGTGTCAGGTGTGGTTAAGTCTGA TGGCCCCAACCTAGCAATGGAGCTAGATGTTCACCTCAATGAACCTACCCCCGCCAGCTGCTTCAAGCATAGCCTCACTGCCTACAGCCAGCTGTAAGGCTTGGCAGCTGTTCAATCAACTTACGGGTCAAAGGTAGTGCAATGTCGCGATCCTAG ATAAGTCCGTATCGTTCCAATTTTGTTACACAATTAGAAATTGGGCTGTCAAGCTCACTTTAAGAACCGACCACAAGGGCTGGCGCATATTGGTTACGAACCTCGGTGTTTAGCGCCTCCCATTCATTCTACGCAGTGGGACTAGTTCGTAGTACG CCTGGAGTACACGTTACGTGTCAAACGTATTTAGGGAAGACGCTCTGCCGTGATCTCGAAAGTTCACTCAACCAAGTTACGATAGCGTAGCTGATGCACGCACTTTCGTACTATCCAGGGCGGCGGCCTGCGCTAGTCACCTAGCTCATTGCCCAA CCCAGTGCCGGGTCGATACCGCCCCTGTTAAAGACTTCTCATAATTAAGCCTCCTTAGCAAGCCGAGATACGGACACCGGGCCCTTGCGGGCCTGCAAGTGGCACGTACCCACAGCGCAGCTAGACCTCTTCATTGTAAGAATATCTTCAATCGAC TCATCCGATCGTCACAACTGTTCATCTCCGGGCCATTTGGTTATAGTTTATTCAATCCACACCACCTCACCAAGGGAATGAAAATACCGAGACGGATGCGGAATGATGCTATGGGCAATTTCTAGCATAGTGATTGGGATGCTATATGGGAGGCGA TCACACGGTGCCTGCGTAGGGTTAGGGCAGCGGGTCTGCCCTACACAAAGGTAGACAACATGGATCTCAACCGGTCCCCACTTAGACTAGAACTCTCCGCCCGAGCCAGTTTCAATCAACGAGATTTTTCCGGACCTATTCTCGCCTCTAGTGTTA AGCTTCCCTGACCCTCGGTGGATAGTACCGTTAATACTCGGCGTCTTGTGCGACGGTATTGGTTCAGTCTACCGTTCTCGGGGCTGTACCCACCTAATTCGTGCCTGAGTCTCCCAATCAGCCTGGTTCTGGCCGACAACAGCAACTGTGCGCGCG ACACGGTTAGCTGATCAGTCTCAAAACACAGCACGTCACAAGGGCGAACAGCAGAGCGAGGGAATCGAACGCACATTGGTTTGTCTGAACCCCGCCCCTTCATTCGACTGACGAGTACGACCGAACTGTAACCTTAGGTGCTTAGGCCCATAAACA GCTTTAAATTCAGGCAGGAGGCATGTGACGATTTTAAAGCGACTATTTGCGCACACGTGCTTTAAGCAGTATTTTTCAATCCACGGTTGATGCGTCCAAGTCAGTTTCAAGACCTAGAAAATATGGATGATATTCCACCACAGCCTCCGCGTGATA CGAACCCTTTCGCCAATATATGACGACCACCGAAATAGTTCAGTCGACCCATCTACGCTCATGTGCGGAGACCAGCGCTGGTGACCTCTGAATTATGGAGACTCTAATCGCGACGACATCCCCAAGAATATGTCATGAAGGGCCAGGTCAATGCAA TCACTGCGGGCCACTTCAGTCGACAGCGTGTATATTGCCTAGATTCCTCTTGAAGCACTTGAGATAAATAACGGATCTTATCACCGGATATTTTGCTATCCACCTCATACCTTTAATTAAACCTTTTATCGGAACGCAGGGTGTACTCGGACTAGG TTCTGACTCTACGAGGGGGACCGCTTGGCTTTCCTTTGAGGCGGTTACATTGTCATTTTGTATTCAGTCTACCCGGGGGCCGACTAGTGGAGGCGGTGGCACATTCGGGCCATTGTCTACACAGTTGCTCGGGAAGCGGTGTTCAGATCCTGTTTA CTGGTTACACTCCGTTGTCGGTCCTAGGCTGTGATGTCTAGCCCCTGAATCAAAGCATGCCAGGGTTTCCAAGAGTCGGTACGTGTGCACCGTTGCCTTAAGTCTTAGTTCACGCCCTGGCTGAGAGTGTGCGCTATTGCAACTGGTTCATTCAAC CTAAGATTCCTTGGCCGTACATAAAGGCGCTGATCGACGCTGCCTTGGTGCGTACGGTGAGAAGCGTGCTTAATTTCAGATATCCCCTATCTGTCTGGATGTTATTCTCATTCATTCAACGTCGGTGGGCATCTCGTATGTTCAGATGGGATAGAC TCCACTCGTTCCATCAGCCGGTGGCCCTCGAGAGCTTTAGTGCTGACGAGAGGTGTCACAGGTTCAATCCACTCCTGTTTGGATTGTACAAGGGGCCAGGCTGGTTAGAAATGTTAACTCTCGCTACATCTGCGCGGGACGGCGCTAGTCAGTCGC CGGGAGGATCTTAGTCCCAACATTGATGCGGCTATAAGGAAGATAGGGCTGCCGTTTGGCACGTCTCGTTATTCAAAGCAAAATTTTGAGGCCATCCCAATAACCGTACCACAAGACAGAGTTTACTTCTCGGAGGGAAGATGCGCTTCATTCGAC TACGCGCCGAGCACCTGTGGAGTGGGTGTGGGCAAGACCCTAGTTACATGCGGCACGGGATTCTGGAAGTAAACATAGTATGCTCTGACACAACTTGATGACGATAACTATCCACAGAGCAAAATGATTCATCATCTTTTCGGGATTTCAGTCTAC CTCCACTAAGTACGCATATTCCAAACCTCAGACATCATACATGTATAAGTCTTATATGTTAGATAGGTTAAACCTTCAGAGTTTTTAGGCTAACTCTATAGGTTGAGCGGCATTGGACAGACACGTGGAGGCAATTCAGTCTACTCTAGGACCCGA TTGGGCATCCCGCTTCAGAGGGCCTACCCTCTTGGCACTGGGACGTTAAGTTCACTCTACTTAAATGCCCATTGGTGGGATCCTCCTTCTAGAATAACCGAACAAGCATTGTAAGTTATTCTTAGGAAAGGCACGTATGTCCAGTGAAGGGTCATC GAACCGGCCGCGCCTCTGGCAGAGTGCCGCTATCAGGCGCACCCCGCCATCCACATGTGGGATGGAGCTAATGAATGGCAACTAAGTTCACTCAACGTTGATTTCTCGAATGAATCCTGACCTGCTTGCCAGGGCGTGATTTATCATTTGGGTGAT TGGCTACATTAGTAACAGATCGAGGACTAACTGATGATTTCAGTCAACCTTGCCGGAAAGTGCGGGTGACATCCAAGAGTAGCAAGTGTGGACTCACGAGGCTTGAAGGATACTGCCTGGCGATCTTCATATATTGGCAGTCAACCTCAGTTCGTA AATCGAGTGGGGCTATTCCACAAATAAAACAACCATTAAATTGGCCCCGAAATCTAGACACAACCAATATAGAGTTCAGTCTACTACAGTTGGTCTGTTTGGCGTACCCGATCTCCGCCTCATTTAATCTTCGTTTCTAGAGATTTTCTCAATCGC\".split()\n",
    "k = 12\n",
    "t = 25\n",
    "best_motifs = greedy_motif_search(Dna, k, t)\n",
    "for motif in best_motifs:\n",
    "    print(motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f77d7fc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
